### ISSUE TITLE
C SDK AMQP_WS Network Disconnect and Reconnect

### ISSUE THREAD

**Issue by mrhprasanna** [Azure/azure-iot-sdk-c#369]
- **OS and version used:** <Ubuntu 14.04> 

- **SDK version used:** <1.1.29> 

# Description of the issue:

Azure C SDK is taking 4 minutes to identify the Network Disconnection and Network Availability while using AMQP_WS protocol. Let's say an application is started at 10.00 A.M and listing here the behaviour  of C SDK.
10.00 A.M ==> Application started
                 ==> Send some messages and receive confirmation from IOTHUB
10.02 A.M ==> Unplug the ethernet cable
                 ==> Send messages at some regular interval
10.06 A.M ==> Logs will be started printing for transport re connection ( Network disconnection logs are started printing)
                 ==> Connection Status call back will be updated
                 ==> Continue sending messages at regular interval
10.36 A.M ==> Connect the ethernet cable
10.37 A.M ==> Connection status call back is updated
10.38 A.M ==> CONFIRMATION_ERROR for messages sent between 10.02 A.M to 10.06 A.M
                 ==> Send some messages 
(Though Network connection is available, confirmation is not received from IOTHUB for messages)

10.42 A.M ==> CONFIRMATION_ERROR for messages sent started from 10.06 A.M

Though the network connection is available, C SDK is not sending messages immediately and it is taking nearly 4 minutes to tell confirmation after plugging the ethernet cable.

# Code sample exhibiting the issue:
(Each 'enter' key press will send one message )
//============================================================================
// Name        : AzureSample.cpp
// Author      : 
// Version     :
// Copyright   : Your copyright notice
// Description : Hello World in C, Ansi-style
//============================================================================

#include <certs.h>
#include <stdio.h>
#include <stdlib.h>
#include <iostream>

#include "azure_c_shared_utility/platform.h"
#include "azure_c_shared_utility/threadapi.h"
#include "azure_c_shared_utility/crt_abstractions.h"
#include "azureiot/iothub_client.h"
#include "azureiot/iothub_message.h"
#include "azureiot/iothubtransportamqp_websockets.h"
#include "azureiot/iothubtransporthttp.h"
#include <unistd.h>

using namespace std;

static const char* connectionString="";

static int callbackCounter;

void init();
void destroy();
int confirmation = 0;

typedef struct EVENT_INSTANCE_TAG
{
    IOTHUB_MESSAGE_HANDLE messageHandle;
    int messageTrackingId;  // For tracking the messages within the user callback.
} EVENT_INSTANCE;

static IOTHUBMESSAGE_DISPOSITION_RESULT ReceiveMessageCallback(IOTHUB_MESSAGE_HANDLE message, void* userContextCallback)
{
    int* counter = (int*)userContextCallback;
    const unsigned char* buffer = NULL;
	size_t size = 0;
    const char* messageId;
    const char* correlationId;

    // Message properties
    if ((messageId = IoTHubMessage_GetMessageId(message)) == NULL)
    {
        messageId = "<null>";
    }

    if ((correlationId = IoTHubMessage_GetCorrelationId(message)) == NULL)
    {
        correlationId = "<null>";
    }

    // Message content
    IOTHUBMESSAGE_CONTENT_TYPE contentType = IoTHubMessage_GetContentType(message);

	if (contentType == IOTHUBMESSAGE_BYTEARRAY)
	{
		if (IoTHubMessage_GetByteArray(message, &buffer, &size) == IOTHUB_MESSAGE_OK)
		{
            (void)printf("Received Message [%d]\r\n Message ID: %s\r\n Correlation ID: %s\r\n BINARY Data: <<<%.*s>>> & Size=%d\r\n", *counter, messageId, correlationId, (int)size, buffer, (int)size);
		}
		else
		{
			(void)printf("Failed getting the BINARY body of the message received.\r\n");
		}
	}
	else if (contentType == IOTHUBMESSAGE_STRING)
	{
		if ((buffer = (const unsigned char*)IoTHubMessage_GetString(message)) != NULL && (size = strlen((const char*)buffer)) > 0)
		{
            (void)printf("Received Message [%d]\r\n Message ID: %s\r\n Correlation ID: %s\r\n STRING Data: <<<%.*s>>> & Size=%d\r\n", *counter, messageId, correlationId, (int)size, buffer, (int)size);
		}
		else
		{
			(void)printf("Failed getting the STRING body of the message received.\r\n");
		}
	}
	else
	{
		(void)printf("Failed getting the body of the message received (type %i).\r\n", contentType);
	}

    // Retrieve properties from the message
    MAP_HANDLE mapProperties = IoTHubMessage_Properties(message);
    if (mapProperties != NULL)
    {
        const char*const* keys;
        const char*const* values;
        size_t propertyCount = 0;
        if (Map_GetInternals(mapProperties, &keys, &values, &propertyCount) == MAP_OK)
        {
            if (propertyCount > 0)
            {
                size_t index;

                printf(" Message Properties:\r\n");
                for (index = 0; index < propertyCount; index++)
                {
                    printf("\tKey: %s Value: %s\r\n", keys[index], values[index]);
                }
                printf("\r\n");
            }
        }
    }

    (*counter)++;

    /* Some device specific action code goes here... */
    return IOTHUBMESSAGE_ACCEPTED;
}

static void SendConfirmationCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
{
    EVENT_INSTANCE* eventInstance = (EVENT_INSTANCE*)userContextCallback;
    (void)printf("Confirmation[%d] received for message tracking id with result = %s\r\n",callbackCounter,ENUM_TO_STRING(IOTHUB_CLIENT_CONFIRMATION_RESULT, result));
    /* Some device specific action code goes here... */
    callbackCounter++;
    if(result==IOTHUB_CLIENT_CONFIRMATION_OK)
    {
    	confirmation = 1;
	IoTHubMessage_Destroy(eventInstance->messageHandle);
	eventInstance->messageHandle = NULL;
    }
    else
    {
	confirmation = -1;
    }
    
    
}

static char msgText[1024];
static char propText[1024];
IOTHUB_CLIENT_HANDLE iotHubClientHandle;
size_t retryTimeSeconds=0;

static void connectionStatusCallback(IOTHUB_CLIENT_CONNECTION_STATUS callback,
		IOTHUB_CLIENT_CONNECTION_STATUS_REASON reason, void *userContext)
{

	printf("Reason [%d]", reason);
	if(reason==IOTHUB_CLIENT_CONNECTION_NO_NETWORK)
	{
		printf("connection is not available\n");
	}
	else if(reason==IOTHUB_CLIENT_CONNECTION_OK)
	{
		printf("connection is available\n");
	}

}

void init()
{
	int receiveContext = 0;
	if (platform_init() != 0)
		{
			(void)printf("ERROR: failed initializing the platform.\r\n");
		}
//AMQP_Protocol_over_WebSocketsTls HTTP_Protocol
	    else if ((iotHubClientHandle = IoTHubClient_CreateFromConnectionString(connectionString,AMQP_Protocol_over_WebSocketsTls)) == NULL)
	    {
	        (void)printf("ERROR: iotHubClientHandle is NULL!\r\n");
	        platform_deinit();
	    }
	    else
	    {
	    	/* unsigned int timeout = 241000;
	    	            // Because it can poll "after 9 seconds" polls will happen effectively // at ~10 seconds.
	    	            // Note that for scalabilty, the default value of minimumPollingTime
	    	            // is 25 minutes. For more information, see:
	    	            // https://azure.microsoft.com/documentation/articles/iot-hub-devguide/#messaging
	    	 unsigned int minimumPollingTime = 9;
	    	 if (IoTHubClient_SetOption(iotHubClientHandle, "timeout", &timeout) != IOTHUB_CLIENT_OK)
	    	 {
	    		 printf("failure to set option \"timeout\"\r\n");
	    	 }
	    	 if (IoTHubClient_SetOption(iotHubClientHandle, "MinimumPollingTime", &minimumPollingTime) != IOTHUB_CLIENT_OK)
	    	 {
	    	     printf("failure to set option \"MinimumPollingTime\"\r\n");*/
	        // For mbed add the certificate information
	        if (IoTHubClient_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
	        {
	            printf("failure to set option \"TrustedCerts\"\r\n");
	        }

	        /*Setting Message call back, so we can receive Commands. */
	        if (IoTHubClient_SetMessageCallback(iotHubClientHandle, ReceiveMessageCallback, &receiveContext) != IOTHUB_CLIENT_OK)
	        {
	            (void)printf("ERROR: IoTHubClient_SetMessageCallback..........FAILED!\r\n");
	        }
	        if(IoTHubClient_SetRetryPolicy(iotHubClientHandle,IOTHUB_CLIENT_RETRY_POLICY::IOTHUB_CLIENT_RETRY_INTERVAL,retryTimeSeconds)!=IOTHUB_CLIENT_OK)
	        {
	        	printf("Error:Iothubclient_SetRetryPolicy failed..\n");
	        }
	        void *p = NULL;
	        IoTHubClient_SetConnectionStatusCallback(iotHubClientHandle,connectionStatusCallback,p);
	    }
}

void destroy()
{
	IoTHubClient_Destroy(iotHubClientHandle);
	platform_deinit();
}

void iothub_client_sample_amqp_websockets_run(EVENT_INSTANCE* messages)
{

    

    srand((unsigned int)time(NULL));
    double avgWindSpeed = 10.0;
    double minTemperature = 20.0;
    double minHumidity = 60.0;
    double temperature = 0;
    double humidity = 0;

    callbackCounter = 0;

                temperature = minTemperature + (rand() % 10);
                humidity = minHumidity +  (rand() % 20);
                sprintf_s(msgText, sizeof(msgText), "{\"deviceId\":\"myFirstDevice\",\"windSpeed\":%.2f,\"temperature\":%.2f,\"humidity\":%.2f}", avgWindSpeed + (rand() % 4 + 2), temperature, humidity);
                if ((messages->messageHandle = IoTHubMessage_CreateFromByteArray((const unsigned char*)msgText, strlen(msgText))) == NULL)
                {
                    (void)printf("ERROR: iotHubMessageHandle is NULL!\r\n");
                }
                else
                {
                    messages->messageTrackingId = 1;

                    MAP_HANDLE propMap = IoTHubMessage_Properties(messages->messageHandle);
                    (void)sprintf_s(propText, sizeof(propText), temperature > 28 ? "true" : "false");
                    if (Map_AddOrUpdate(propMap, "temperatureAlert", propText) != MAP_OK)
                    {
                        (void)printf("ERROR: Map_AddOrUpdate Failed!\r\n");
                    }
			
		   int a =10;
                    if (IoTHubClient_SendEventAsync(iotHubClientHandle, messages->messageHandle, SendConfirmationCallback, messages) != IOTHUB_CLIENT_OK)
                    {
                        (void)printf("ERROR: IoTHubClient_SendEventAsync..........FAILED!\r\n");
                    }
                    else
                    {
                        (void)printf("IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.\r\n");
                    }
                 }

            /* Wait for Commands. */
           (void)printf("Press any key to exit the application. \r\n");
            (void)getchar();
    }

int main(void)
{
	init();
	EVENT_INSTANCE message[1000];
	int i=0;
	while(true)
	{
	std::cout<<"Slot ID:"<<i<<std::endl;
	iothub_client_sample_amqp_websockets_run(&message[i]);
	i=i+1;
	}
	return EXIT_SUCCESS;
}

# Console log of the issue:

Info: Preparing transport for re-connection
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransportamqp_methods.c Func:iothubtransportamqp_methods_unsubscribe Line:817 unsubscribe called while not subscribed
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:627 Failure: getaddrinfo failure -3.
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/src/wsio.c Func:internal_close Line:150 wsio_close when not open.
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/src/tlsio_openssl.c Func:on_underlying_io_open_complete Line:746 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:1246 Failed opening the underlying I/O.
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/src/uws_client.c Func:uws_client_open_async Line:1472 Opening the underlying IO failed
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/src/wsio.c Func:wsio_open Line:512 Opening the uws instance failed.
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/uamqp/src/saslclientio.c Func:saslclientio_open_async Line:1081 xio_open failed
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/uamqp/src/connection.c Func:connection_open Line:1262 Opening the underlying IO failed
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/uamqp/src/message_receiver.c Func:messagereceiver_open Line:360 Link attach failed
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/uamqp/src/amqp_management.c Func:amqp_management_close Line:887 AMQP management instance not open
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:on_cbs_open_complete Line:153 CBS open failed
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/uamqp/src/amqp_management.c Func:amqp_management_open_async Line:848 Failed opening message receiver
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:create_cbs_handle Line:278 Failed to open the connection with CBS.
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:amqp_connection_create Line:421 amqp_connection_create failed (failed creating the CBS handle)
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:establish_amqp_connection Line:831 Failed establishing connection (failed to create the amqp_connection instance).
Error: Time:Thu Feb  1 12:11:25 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1576 AMQP transport failed to establish connection with service.

Slot ID:26
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
Press any key to exit the application. 
Info: Preparing transport for re-connection
Error: Time:Thu Feb  1 12:11:30 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransportamqp_methods.c Func:iothubtransportamqp_methods_unsubscribe Line:817 unsubscribe called while not subscribed
Reason [5]Error: Time:Thu Feb  1 12:12:17 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:1123 Failed performing DoWork for device 'XXXXXXXXXXXXXX' (device reported state 5; number of previous failures: 0)
Reason [6]connection is available

Slot ID:27
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
Press any key to exit the application. 
Reason [5]Error: Time:Thu Feb  1 12:12:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:1123 Failed performing DoWork for device 'XXXXXXXXXXXXXX' (device reported state 5; number of previous failures: 1)
Reason [6]connection is available
Error: Time:Thu Feb  1 12:13:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:1111 Failed performing DoWork for device 'XXXXXXXXXXXXXXX' (device failed to start or stop within expected timeout)
Error: Time:Thu Feb  1 12:13:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_Device_DoWork Line:1123 Failed performing DoWork for device 'XXXXXXXXXXXXXXX' (device reported state 4; number of previous failures: 2)
Confirmation[0] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[1] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[2] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[3] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[4] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[5] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[6] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[7] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Reason [6]connection is available
Error: Time:Thu Feb  1 12:13:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1593 Device 'XXXXXXXXXXXX' reported a critical failure (events completed sending with failures); connection retry will be triggered.
Info: Preparing transport for re-connection
Error: Time:Thu Feb  1 12:13:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransportamqp_methods.c Func:iothubtransportamqp_methods_unsubscribe Line:817 unsubscribe called while not subscribed
Error: Time:Thu Feb  1 12:13:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:on_cbs_error Line:160 CBS Error occured
Error: Time:Thu Feb  1 12:13:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:on_amqp_connection_state_changed Line:762 amqp_connection was closed unexpectedly; connection retry will be triggered.
Reason [6]connection is available

Slot ID:28
IoTHubClient_SendEventAsync accepted data for transmission to IoT Hub.
Press any key to exit the application. 
Error: Time:Thu Feb  1 12:18:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_connection.c Func:on_cbs_error Line:160 CBS Error occured
Confirmation[0] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[1] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[2] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[3] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[4] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[5] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[6] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[7] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[8] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[9] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[10] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[11] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[12] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[13] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[14] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[15] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[16] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[17] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Confirmation[18] received for message tracking id with result = IOTHUB_CLIENT_CONFIRMATION_ERROR
Error: Time:Thu Feb  1 12:18:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:IoTHubTransport_AMQP_Common_DoWork Line:1593 Device 'XXXXXXXXXXX' reported a critical failure (events completed sending with failures); connection retry will be triggered.
Info: Preparing transport for re-connection
Error: Time:Thu Feb  1 12:18:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransportamqp_methods.c Func:iothubtransportamqp_methods_unsubscribe Line:817 unsubscribe called while not subscribed
Error: Time:Thu Feb  1 12:18:47 2018 File:/home/administrator/azure-iot-sdk-c/iothub_client/src/iothubtransport_amqp_common.c Func:on_amqp_connection_state_changed Line:762 amqp_connection was closed unexpectedly; connection retry will be triggered.
Reason [4]connection is not available
Reason [6]connection is available

**Comment by ewertons** [comment]
Hi @mrhprasanna,
if the OS doesn't raise a failure the SDK does not have an immediate way to inform if there is a failure.
The SDK than resorts to its timeout limits for keep-alive and message timeout.
The keep alives are sent to the socket every 60 seconds, and if no response is received within 4 minutes the SDK assumes there is a network failure, and triggers the whole list of failures you see in the logs.

You can set these timeouts in the SDK by using the following options:
size_t keepAliveFrequencySecs = 60;
IoTHubClient_SetOption(iothubClientHandle, OPTION_SERVICE_SIDE_KEEP_ALIVE_FREQ_SECS, &keepAliveFrequencySecs);

One question: As of now, after waiting past 4 minutes do you see the SDK reconnecting to the IoT Hub at all?

**Comment by jspaith** [comment]
Another point @mrhprasanna  - your SDK version is 1.1.29, which I believe is corresponding to our 12/15/2017 release, correct?  It's tricky here we have the SDK version in https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/inc/iothub_client_version.h which had a 1.1.29, but we also have an SDK version corresponding to 1/29 as in January 29th https://github.com/Azure/azure-iot-sdk-c/releases/tag/2018-01-29 so want to make sure no ambiguity)

Since the 12/15 release, we have made further improvements in our reconnection logic.  In particular we're better about detecting network disconnections from the OS - https://github.com/Azure/azure-c-shared-utility/commit/c9de11eb50f5f8b11a43f0f515c586b4d91154f0#diff-daf49d508fcb5ec213199d0e0bafbe87.  

So before going too deep could you upgrade your SDK and see if the issue still repros?  Thanks

**Comment by mrhprasanna** [comment]
Hi @ewertons  and @jspaith 
                        I have used the option suggested by you and now am able to get the logs immediately with the latest release version. 
                      But the issue is still reproducible and C SDK is sending "Confirmation" after 4 minutes of connecting the physical cable.

**Comment by ewertons** [comment]
@mrhprasanna , could you share the full logs please? from beginning to end.

Also: As of now, after waiting past 4 minutes do you see the SDK reconnecting to the IoT Hub at all?

**Comment by mrhprasanna** [comment]
[Azure_Network_Log.txt](https://github.com/Azure/azure-iot-sdk-c/files/1694374/Azure_Network_Log.txt)

@ewertons , I have provided here you the enough logs that i have observed in my machine, 

After waiting past 4 minutes, I can see the SDK reconnecting to the IoT Hub.

**Comment by ewertons** [comment]
Thanks @mrhprasanna,

based on the logs you shared the SDK is retrying to connect as expected, but the system is not having connectivity.

If you look at the logs you will see these traces:

Error: Time:Mon Feb  5 12:35:49 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:627 Failure: getaddrinfo failure -3.
...
Error: Time:Mon Feb  5 13:08:39 2018 File:/home/administrator/azure-iot-sdk-c/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:627 Failure: getaddrinfo failure -3.

That failure comes from getaddrinfo, outside the SDK.
To have more details to undestand what is happening, you could:
1. Start a constant ping in parallel with the SDK sample;
2. take note of the time you disconnect and reconnect the cable
3. Take note of the time the pings start failing, and when they start working again;
4. Compare the time the pings start working again and the time the SDK reconnects (this is the real delay we should make sure it is as short as possible).

Based on the timestamps I see you set the retry policy to INTERVAL.
Could you also confirm you adjusted the OPTION_SERVICE_SIDE_KEEP_ALIVE_FREQ_SECS (see comment above).

**Comment by ewertons** [comment]
Do you have the logs?

**Comment by ewertons** [comment]
And also the list of events with timestamps?
Note that in our tests the SDK reconnects right after the network is back, that's why we need your data.

**Comment by mrhprasanna** [comment]
@ewertons , what type of data do you actually need since i actually given you logs in my first query itself.

**Comment by markrad** [comment]
The return code -3 from getaddrinfo is a "temporary failure in name resolution" though it doesn't look very temporary from the log. Perhaps you should try running nslookup <yourhub>.azure-devices.net after you reconnect the cable and see if that is able to resolve the name.

_Mark Radbourne MSFT_

**Comment by mrhprasanna** [comment]
@markrad I will find the result and update you soon

**Comment by ewertons** [comment]
Hi @mrhprasanna ,
we will close this issue for now but if you would like to follow up please feel free to reopen it.
Thanks,
Azure IoT SDK Team.

**Comment by nitu011** [comment]
Hi ,
I am also facing the same issue , iot disconnect message is coming after 4 minutes only. is there any way I can resolve this. I checked the code , fix is already present in my source code. The option you suggested is throwing following error:

/home/nitu/develop/gateway-app/sentience2.cloud_connector.cpp/src/message/IoTHubMessagingClient.cpp:239:56: error: 'OPTION_SERVICE_SIDE_KEEP_ALIVE_FREQ_SECS' was not declared in this scope
   ::IoTHubClient_SetOption(this->iothub_client_handle, OPTION_SERVICE_SIDE_KEEP_ALIVE_FREQ_SECS, &keepAliveFrequencySecs);


Please help me to fix this issue.. we are loosing critical events

**Comment by jspaith** [comment]
@nitu011 - our team typically doesn't track closed GitHub issues.  Could you please open up a new issue and reference back to this one?

In the new issue creation (https://github.com/Azure/azure-iot-sdk-c/issues/new?assignees=&template=bug-report.md&title=) there's a number of questions we ask up front so if you could answer as many of those as possible when creating your entry it'll help expedite the investigation.
